File: ai/workflows/transferWorkflow.js
import { runTransferGraph } from './transfer/transferWorkflow.js';
import { getWindowToolAction, getWindowToolReply } from '../assistant/responseWrappers.js';
import {
  createExecutedWorkflowResponse,
  createWorkflowResponse
} from '../contracts/assistantResponseContract.js';

export const runTransferWorkflow = async ({ state, services, createChatCompletion, abortSignal }) => {
  const result = await runTransferGraph({
    userInput: state.userInput,
    userLanguage: state.session.userLanguage,
    userId: state.session.userId,
    transferState: state.transfer.nextTransferState,
    semanticIntent: state.intent?.intent || state.intent?.detectedIntent,
    transferPayload: state.intent?.transferPayload || null,
    correction: state.intent?.correction || null,
    services,
    createChatCompletion,
    abortSignal
  });
  const nextTransferState = result.nextTransferState || state.transfer.nextTransferState;
  const activeWorkflow = nextTransferState?.phase && nextTransferState.phase !== 'idle'
    ? 'transfer_workflow'
    : 'unknown';

  if (!result.handled) {
    const formResult = await services.transactionService.openTransferForm({ userId: state.session.userId });
    const workflowResponse = createExecutedWorkflowResponse({
      message: getWindowToolReply('open_money_transfer_inline', state.session.userLanguage),
      action: getWindowToolAction('open_money_transfer_inline', formResult),
      operation: 'open_money_transfer_inline',
      result: formResult
    });

    return {
      ...state,
      workflow: { ...state.workflow, activeWorkflow, currentPhase: 'Return Response with Suggestions' },
      execution: workflowResponse.execution,
      workflowResponse,
      ui: {
        ...state.ui,
        message: workflowResponse.message,
        action: getWindowToolAction('open_money_transfer_inline', formResult)
      }
    };
  }

  const workflowResponse = createWorkflowResponse({
    message: result.reply || '',
    action: result.action || null,
    nextConversationState: nextTransferState,
    execution: {
      executed: Boolean(result.handled),
      operation: 'transfer_money',
      result
    }
  });

  return {
    ...state,
    workflow: { ...state.workflow, activeWorkflow, currentPhase: 'Return Response with Suggestions' },
    transfer: {
      ...state.transfer,
      receiverEmail: nextTransferState?.receiverEmail || '',
      amount: nextTransferState?.amount ?? null,
      description: nextTransferState?.description || '',
      confirmationRequired: Boolean(nextTransferState?.riskConfirmationAsked),
      phase: nextTransferState?.phase || 'idle',
      lastValidationError: nextTransferState?.lastValidationError || null,
      nextTransferState
    },
    execution: workflowResponse.execution,
    workflowResponse,
    ui: {
      ...state.ui,
      message: workflowResponse.message,
      action: result.action || null
    }
  };
};

File: ai/workflows/transfer_workflow.py
"""
Transfer workflow — port of transferWorkflow.js (the outer one).
"""

from ai.workflows.transfer.transfer_state_machine import run_transfer_state_machine, build_next_transfer_state
from ai.assistant.response_wrappers import get_window_tool_action, get_window_tool_reply
from ai.contracts.assistant_response_contract import create_executed_workflow_response, create_workflow_response


async def run_transfer_workflow(
    *, state: dict, services: dict, create_chat_completion=None, abort_signal=None
) -> dict:
    session = state.get("session") or {}
    intent = state.get("intent") or {}
    transfer = state.get("transfer") or {}

    result = await run_transfer_state_machine(
        user_input=state.get("userInput", ""),
        user_language=session.get("userLanguage", "en"),
        user_id=session.get("userId"),
        transfer_state=transfer.get("nextTransferState"),
        semantic_intent=intent.get("intent") or intent.get("detectedIntent") or "unknown",
        transfer_payload=intent.get("transferPayload"),
        correction=intent.get("correction"),
        services=services,
        create_chat_completion=create_chat_completion,
        abort_signal=abort_signal,
    )

    next_transfer_state = build_next_transfer_state(result)
    phase = (next_transfer_state or {}).get("phase", "idle")
    active_workflow = "transfer_workflow" if phase and phase != "idle" else "unknown"

    if not result.get("handled"):
        form_result = await (services or {}).get("transactionService").open_transfer_form(
            user_id=session.get("userId")
        )
        reply = get_window_tool_reply("open_money_transfer_inline", session.get("userLanguage", "en"))
        action = get_window_tool_action("open_money_transfer_inline", form_result)
        workflow_response = create_executed_workflow_response(
            message=reply,
            action=action,
            operation="open_money_transfer_inline",
            result=form_result,
        )
        return {
            **state,
            "workflow": {**(state.get("workflow") or {}), "activeWorkflow": active_workflow, "currentPhase": "Return Response with Suggestions"},
            "execution": workflow_response["execution"],
            "workflowResponse": workflow_response,
            "ui": {**(state.get("ui") or {}), "message": reply, "action": action},
        }

    workflow_response = create_workflow_response(
        message=result.get("reply", ""),
        action=result.get("action"),
        next_conversation_state=next_transfer_state,
        execution={
            "executed": bool(result.get("handled")),
            "operation": "transfer_money",
            "result": result,
        },
    )

    return {
        **state,
        "workflow": {**(state.get("workflow") or {}), "activeWorkflow": active_workflow, "currentPhase": "Return Response with Suggestions"},
        "transfer": {
            **(state.get("transfer") or {}),
            "receiverEmail": (next_transfer_state or {}).get("receiverEmail", ""),
            "amount": (next_transfer_state or {}).get("amount"),
            "description": (next_transfer_state or {}).get("description", ""),
            "confirmationRequired": bool((next_transfer_state or {}).get("riskConfirmationAsked")),
            "phase": (next_transfer_state or {}).get("phase", "idle"),
            "lastValidationError": (next_transfer_state or {}).get("lastValidationError"),
            "nextTransferState": next_transfer_state,
        },
        "execution": workflow_response["execution"],
        "workflowResponse": workflow_response,
        "ui": {
            **(state.get("ui") or {}),
            "message": workflow_response["message"],
            "action": result.get("action"),
        },
    }
